From 7ae1792bd982787b46bcc017cbcb9cb7411d9a35 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 06 Dec 2021 06:12:11 +0200
Subject: [PATCH] [SW-0] ppboot: set privileged mode to start of ppboot

Currently, ARCs are set to privileged mode as part of ppboot security
init function. However till that point ARCs mode is changed to user
mode due to reset and it tries to access registers which are set to
privilege/secure mode and not been reset as part of reset. For example
CPU_PLL registers are marked as privileged and ppboot access it as part of
udelay init where ARCs are still in user mode. To resolve this issue,
set ARCs to privilege mode at very start of ppboot.

Change-Id: I976cf09141cd4a368308a589909d4bbb43542611
---

diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index 32824c5..cb43fff 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -262,6 +262,7 @@
 }
 
 __weak void ppboot_security_init(void) {}
+__weak void ppboot_secure_arc(void) {}
 
 void ppboot_main(void)
 {
@@ -269,6 +270,9 @@
 	bool ret;
 
 	ppboot_init_stack_monitor();
+	/* Set ARCs to privileged mode */
+	ppboot_secure_arc();
+
 	/* Initialize internal if variables */
 	cpu_boot_if_init();
 	hl_plat_early_init();
diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index a345421..d6f66d4 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -416,7 +416,8 @@
 	  .data_size = ARRAY_SIZE(gaudi2_protbits_data0)
 	},
 };
-void preboot_secure_arc(void)
+
+void ppboot_secure_arc(void)
 {
 	/* enable security and privilege mode for arc0 and arc1 */
 	regmap_writel(EN_ARC_SEC_PRIV,
@@ -864,8 +865,6 @@
 
 void ppboot_security_init(void)
 {
-	preboot_secure_arc();
-
 	preboot_secure_elbi();
 
 	/* Initialize only if security is enabled */
diff --git a/include/preboot.h b/include/preboot.h
index 4cd55f7..1add4e0 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -244,6 +244,7 @@
 void ppboot_security_init(void);
 
 void preboot_main(void);
+void ppboot_secure_arc(void);
 
 #endif /* __ASSEMBLY__ */
 
